src/core/server/deprecations/types.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { IScopedClusterClient, SavedObjectsClientContract } from '../../server';
+
+type MaybePromise<T> = T | Promise<T>;
+
+export interface PluginDeprecationDetails extends DeprecationsDetails {
+  pluginId: string;
+}
+
+export interface DeprecationsDetails {
+  message: string;
+  level: 'warning' | 'critical';
+  documentationUrl?: string;
+  correctiveActions: {
+    api?: {
+      path: string;
+      method: 'POST' | 'PUT';
```

Comment:
I believe each domain should implement a `POST/PUT` route for their plugin to handle their deprecations since a lot of those corrective actions might include multi-steps and some validations before complete. here we only allow 1 api call to keep it simple for the user and upgrade assistant to handle the automatic correctiveActions.

Reply:
Great point, sounds like a good defensive decision.